# Aircraft Capacity planning

In [1]:
#!pip install pulp

- Assumptions: Each cargo can be splitted into any proportion, between compartments such that weight capacity maintains balance of the plane. 
- Let xij>=0 be number of tones of cargo i(1:4) put into compartments j(1:3).
- Objective is to maximize the profit by determining how much of each cargo should be distributed among compartments

In [1]:
#Initialize the problem
from pulp import *
ac_max=LpProblem("Aircraft Capacity planning",LpMaximize)

C:\Users\Admin\anaconda3\envs\graphs\Lib\site-packages\pulp\pulp.py:1298: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


In [2]:
#Decision variables
x11=LpVariable("x11",lowBound=0) #tons of cargo 1 in compartment 1
x12=LpVariable("x12",lowBound=0)
x13=LpVariable("x13",lowBound=0)
x21=LpVariable("x21",lowBound=0)
x22=LpVariable("x22",lowBound=0)
x23=LpVariable("x23",lowBound=0)
x31=LpVariable("x31",lowBound=0)
x32=LpVariable("x32",lowBound=0)
x33=LpVariable("x33",lowBound=0)
x41=LpVariable("x41",lowBound=0) #Tons of cargo 4 in compartment 1
x42=LpVariable("x42",lowBound=0)
x43=LpVariable("x43",lowBound=0)

In [3]:
#Profit per ton * weight of cargo in all the compartment
ac_max +=310*(x11+x12+x13)+380*(x21+x22+x23)+350*(x31+x32+x33)+285*(x41+x42+x43)

In [4]:
ac_max

Aircraft_Capacity_planning:
MAXIMIZE
310*x11 + 310*x12 + 310*x13 + 380*x21 + 380*x22 + 380*x23 + 350*x31 + 350*x32 + 350*x33 + 285*x41 + 285*x42 + 285*x43 + 0
VARIABLES
x11 Continuous
x12 Continuous
x13 Continuous
x21 Continuous
x22 Continuous
x23 Continuous
x31 Continuous
x32 Continuous
x33 Continuous
x41 Continuous
x42 Continuous
x43 Continuous

In [5]:
#Constraints on weight of each cargo in three compartments (Cargo)
ac_max +=x11+x12+x13<=18 #Cargo 1
ac_max +=x21+x22+x23<=16 #Cargo 2
ac_max +=x31+x32+x33<=23 #Cargo 3
ac_max +=x41+x42+x43<=12 #Cargo 4 weight restriction

In [6]:
#Constraints on weight capacity of each compartments (Compartments)
ac_max +=x11+x21+x31+x41<=10 #Weight in Front compartment 
ac_max +=x12+x22+x32+x42<=16 #Center Compartment 
ac_max +=x13+x23+x33+x43<=8  #Rear COmpartment 

In [7]:
#Constraints on Space capacity of each compartment
ac_max +=480*x11+650*x21+580*x31+390*x41<=6800 #Front
ac_max +=480*x12+650*x22+580*x32+390*x42<=8700 #Center
ac_max +=480*x13+650*x23+580*x33+390*x43<=5300 #Rear

In [8]:
#Maintain the balance of the plane by keeping same weight
c1=(x11+x21+x31+x41)/10
c2=(x12+x22+x32+x42)/16
c3=(x13+x23+x33+x43)/8
ac_max +=((x11+x21+x31+x41+x12+x22+x32+x42+x13+x23+x33+x43)/(10+16+8))>=0
ac_max

Aircraft_Capacity_planning:
MAXIMIZE
310*x11 + 310*x12 + 310*x13 + 380*x21 + 380*x22 + 380*x23 + 350*x31 + 350*x32 + 350*x33 + 285*x41 + 285*x42 + 285*x43 + 0
SUBJECT TO
_C1: x11 + x12 + x13 <= 18

_C2: x21 + x22 + x23 <= 16

_C3: x31 + x32 + x33 <= 23

_C4: x41 + x42 + x43 <= 12

_C5: x11 + x21 + x31 + x41 <= 10

_C6: x12 + x22 + x32 + x42 <= 16

_C7: x13 + x23 + x33 + x43 <= 8

_C8: 480 x11 + 650 x21 + 580 x31 + 390 x41 <= 6800

_C9: 480 x12 + 650 x22 + 580 x32 + 390 x42 <= 8700

_C10: 480 x13 + 650 x23 + 580 x33 + 390 x43 <= 5300

_C11: 0.0294117647059 x11 + 0.0294117647059 x12 + 0.0294117647059 x13
 + 0.0294117647059 x21 + 0.0294117647059 x22 + 0.0294117647059 x23
 + 0.0294117647059 x31 + 0.0294117647059 x32 + 0.0294117647059 x33
 + 0.0294117647059 x41 + 0.0294117647059 x42 + 0.0294117647059 x43 >= 0

VARIABLES
x11 Continuous
x12 Continuous
x13 Continuous
x21 Continuous
x22 Continuous
x23 Continuous
x31 Continuous
x32 Continuous
x33 Continuous
x41 Continuous
x42 Continuous
x43 Cont

In [9]:
#Solving LPP
result=ac_max.solve()
result

1

In [10]:
LpStatus[result]

'Optimal'

In [11]:
ac_max.variables()

[x11, x12, x13, x21, x22, x23, x31, x32, x33, x41, x42, x43]

In [12]:
#Optimal profit and decision variable values
print("Objective Function Max Z =",value(ac_max.objective)) 
print("Optimality Status:",LpStatus[result])
print("Decision Variables")
for v in ac_max.variables():
    print("{}={}".format(v.name,v.varValue))


Objective Function Max Z = 12181.578806
Optimality Status: Optimal
Decision Variables
x11=0.0
x12=0.0
x13=0.0
x21=10.0
x22=0.0
x23=6.0
x31=0.0
x32=12.947368
x33=2.0
x41=0.0
x42=3.0526316
x43=0.0


In [13]:
ac_max.constraints.items()

odict_items([('_C1', 1*x11 + 1*x12 + 1*x13 + -18 <= 0), ('_C2', 1*x21 + 1*x22 + 1*x23 + -16 <= 0), ('_C3', 1*x31 + 1*x32 + 1*x33 + -23 <= 0), ('_C4', 1*x41 + 1*x42 + 1*x43 + -12 <= 0), ('_C5', 1*x11 + 1*x21 + 1*x31 + 1*x41 + -10 <= 0), ('_C6', 1*x12 + 1*x22 + 1*x32 + 1*x42 + -16 <= 0), ('_C7', 1*x13 + 1*x23 + 1*x33 + 1*x43 + -8 <= 0), ('_C8', 480*x11 + 650*x21 + 580*x31 + 390*x41 + -6800 <= 0), ('_C9', 480*x12 + 650*x22 + 580*x32 + 390*x42 + -8700 <= 0), ('_C10', 480*x13 + 650*x23 + 580*x33 + 390*x43 + -5300 <= 0), ('_C11', 0.029411764705882353*x11 + 0.029411764705882353*x12 + 0.029411764705882353*x13 + 0.029411764705882353*x21 + 0.029411764705882353*x22 + 0.029411764705882353*x23 + 0.029411764705882353*x31 + 0.029411764705882353*x32 + 0.029411764705882353*x33 + 0.029411764705882353*x41 + 0.029411764705882353*x42 + 0.029411764705882353*x43 + 0.0 >= 0)])

In [14]:
#Sensitivity Analysis
#Shadow price and slack
import pandas as pd
shadow = [{'name':name, 'shadow price':c.pi,'slack': c.slack} for name, c in ac_max.constraints.items()]
pd.DataFrame(shadow).style.background_gradient(cmap='YlGnBu')

### Shadow Price - How much the objective function will improve if we increase the right-hand side (RHS) of a constraint by one unit.
- With ac_max +=x21+x22+x23<=16 #Cargo 2. Optimal Value is 12181.58
- For a unit increase in #Cargo 2- ac_max +=x21+x22+x23<=17. Optimal Value is 12211.58.

### Slack - Unused RHS.
- Unused Space capacity in compartments cubic meters
    - Front - 300 -C8
    - Rear  - 170 -C10